lib/unbeatable_player.ex

```diff
@@ -1,103 +1,75 @@
+
+  def negamax(current_board, depth, alpha, beta) do
+    current_board
+    |> finish_calculation?(depth)
+    |> result(current_board, depth, {alpha, beta})
   end
 
-  def play_move(board, computer_mark, current_mark, depth, alpha, beta) do
+  defp result(:finish, board, depth, _), do: %{:best_score => score(board, depth), :best_move => @move_placehoder}
+  defp result(:continue, board, depth, {alpha, beta}) do
     board
     |> Board.available_positions
-    |> Enum.reduce_while([initialize_score(current_mark, computer_mark), @move_placeholder],
-       fn(next_move), [existing_score, existing_best_move] ->
-      board
-      |> Board.place_mark(next_move)
-      |> get_next_board
-      |> minimax(computer_mark, switch(current_mark), depth - 1, alpha, beta)
-      |> get_new_score
-      |> find_best_score(existing_score, current_mark, computer_mark, next_move, existing_best_move)
-      |> prune(computer_mark, current_mark, alpha, beta)
+    |> Enum.reduce(accumulator(alpha, beta), fn(next_move, result_accumulator) ->
+      check_result(board, next_move, result_accumulator, depth)
     end)
   end
 
-  defp finished?(board, depth) do
-    if depth == 0 || Board.game_over?(board), do: {:finished, board}, else: {:not_finished, board}
-  end
-
-  defp result({:finished, board}, computer_mark, _, depth, _,_) do
-    [score_for_move(board, computer_mark, depth), -1]
-  end
-  defp result({:not_finished, board}, computer_mark, current_mark, depth, alpha, beta) do
-    play_move(board, computer_mark, current_mark, depth, alpha, beta)
-  end
-
-  defp get_next_board({_, next_board}), do: next_board
-
-  defp get_new_score([new_score, _]), do: new_score
-
-  defp find_best_score(new_score, existing_score, current_mark, computer_mark, next_move, existing_best_move) do
-    if new_best_score?(current_mark, computer_mark, new_score, existing_score) do
-      {new_score, next_move}
+  defp check_result(board, move, current_best, depth) do
+    if current_best[:alpha] >= current_best[:beta] do
+      current_best
     else
-      {existing_score, existing_best_move}
-    end
-  end
-
-  defp prune({best_score, best_move}, computer_mark, current_mark, alpha, beta) do
-    new_alpha = update_alpha(alpha, best_score, current_mark, computer_mark)
-    new_beta = update_beta(beta, best_score, current_mark, computer_mark)
-
-    if new_alpha >= new_beta do
-      {:halt, [best_score, best_move]}
-    else
-      {:cont, [best_score, best_move]}
+      board
+      |> Board.place_mark(move)
+      |> negamax(depth - 1, -current_best[:beta], -current_best[:alpha])
+      |> negate_score
+      |> better_score?(current_best)
+      |> update_score(move)
     end
   end
 
-  defp update_alpha(alpha, best_score, current_mark, computer_mark) do
-    if current_mark == computer_mark, do: Enum.max([alpha, best_score]), else: alpha
+  defp update_score({false, _, current_best}, _), do: current_best
+  defp update_score({true, new_result, current_best}, move) do
```

Comment:
I don't really like the name 'update score'.
It feels so OO/non-functional in the sense that we never update anything. We create new copies of stuff. In this case, you are putting in two things and picking the best.
I think there is a better name out there. I'd also not split `update_score` and `better_score?` into separate functions...